### ISSUE TITLE
Is it possible to get the HTTP status code of Amplify.API.get() (i.e.)

### ISSUE THREAD

**Issue by blattr** [aws-amplify/amplify-flutter#531]
**Describe the bug**
All status code above 299 and 304 (as mentioned in [here](https://docs.amplify.aws/lib/restapi/fetch/q/platform/js)) are internally "rated" as an error and therefore thrown as an exception (see below). In my opinion it should be possible to handle all status codes explicitly because not all of them are a result of failed execution (i.e.404 / no data).

Could you please at least give me a tip where I can get the status code?

This is the code (just the snippet):

```dart
try {
  RestOptions options = RestOptions(apiName: "api", path: "/users/123");
  RestOperation restOperation = Amplify.API.get(restOptions: options);
  RestResponse response = await restOperation.response;
  print(String.fromCharCodes(response.data));
} on ApiException catch (e) {
  print(e);
}
```

... and this the exception (because user 123 doesn't exist).

```text
ApiException(message: The HTTP response status code is [404]., recoverySuggestion: The metadata associated with the response is contained in the HTTPURLResponse. For more information on HTTP status codes, take a look at https://en.wikipedia.org/wiki/List_of_HTTP_status_codes, underlyingException: null)
```

**Platform**

[] Android
[x] iOS

**Comment by sivakreddyc** [comment]
any work around. even i'm facing the same issue
When i try directly from AWS console API test it's successful, trying from flutter app it's showing this message

**Comment by blattr** [comment]
This is a really nasty workaround but it works for me (atm). Please don't blame me - I know, it highly depends on the underlying error message :)

Basically, I try to extract the status code from the error message of the thrown APIException:
```dart
final RegExp statusPattern = RegExp("\\[(\\d{3})\\]");

int _tryToExtractStatusCodeFromMessage(String message) {
  RegExpMatch match = statusPattern.firstMatch(message);
  if (match != null) {
    return int.parse(match.group(1));
  }
  return -1;
}
```

.... also, I am wrapping the HTTP result for the calling code in a vary basic way:




```dart
class HttpResponse {
  final HttpResponseState state;
  final Uint8List bodyBytes;

  HttpResponse(this.state, [this.bodyBytes]);
}

enum HttpResponseState { OK, NO_DATA, ERROR }
```

... putting all parts together looks like this:

```dart
Future<HttpResponse> doGet(String path,
      [Map<String, String> queryParameters]) async {
    try {
      Stopwatch stopwatch = Stopwatch()..start();
      RestOptions options = RestOptions(
          apiName: "api", path: path, queryParameters: queryParameters);
      RestOperation restOperation = Amplify.API.get(restOptions: options);
      RestResponse response = await restOperation.response;
      _logger.info(
          "Successfully performed a GET on resource $path in ${stopwatch.elapsedMilliseconds}ms");
      return HttpResponse(HttpResponseState.OK, response.data);
    } on ApiException catch (e) {
      int statusCode = _tryToExtractStatusCodeFromMessage(e.message);
      if (statusCode == HttpStatus.notFound) {
        _logger.info("GET on resource $path returned no data");
        return HttpResponse(HttpResponseState.NO_DATA);
      } else {
        _logger
            .info("GET on resource $path failed with status code $statusCode");
        return HttpResponse(HttpResponseState.ERROR);
      }
    }
  }
```

I hope it helps!

Best regards
Roman

**Comment by sivakreddyc** [comment]
When i set the Authorization type to none it is working fine and setting authorization to IAM type is not working.

raised it is stack overflow: https://stackoverflow.com/questions/67636955/flutter-iam-authorisation-is-throwing-403-with-amplify-api-post-request

**Comment by ragingsquirrel3** [comment]
@blattr to answer your question: no, there is not currently a way to access HTTP status code on the response other than parsing the error message (sorry to put you through that). I was investigating a fix and it might be possible to add an `httpStatusCode` integer property to the `ApiException` class. It would be something like

```dart
  try {
      // ...
      RestOperation restOperation = Amplify.API.get(restOptions: options); // request which fails
      RestResponse response = await restOperation.response
      // ... todo stuff when it works
    } on ApiException catch (e) {
      print("woops... request failed");
      print(e.httpStatusCode);
    }
```

Would that be helpful in this case?

**Comment by ragingsquirrel3** [comment]
@sivakreddyc if I understand correctly, your issue is not necessarily getting the http status code of some error, but that you are getting an error when accessing the API in this case and want to not have an error. Did you enable unauthenticated access to your API when configuring?

**Comment by sivakreddyc** [comment]
@ragingsquirrel3 
No only authenticated users can access, where as get request to another IAM auth enabled API is working fine

**Comment by blattr** [comment]
> @blattr to answer your question: no, there is not currently a way to access HTTP status code on the response other than parsing the error message (sorry to put you through that). I was investigating a fix and it might be possible to add an `httpStatusCode` integer property to the `ApiException` class. It would be something like
> 
> ```dart
>   try {
>       // ...
>       RestOperation restOperation = Amplify.API.get(restOptions: options); // request which fails
>       RestResponse response = await restOperation.response
>       // ... todo stuff when it works
>     } on ApiException catch (e) {
>       print("woops... request failed");
>       print(e.httpStatusCode);
>     }
> ```
> 
> Would that be helpful in this case?

@ragingsquirrel3 IMO that would be the perfect solution :)

**Comment by ragingsquirrel3** [comment]
@blattr a PR has been merged implemented the `e.httpStatusCode` and will be available when the next release goes out. This issue will be closed once that happens.

**Comment by blattr** [comment]
> @blattr a PR has been merged implemented the `e.httpStatusCode` and will be available when the next release goes out. This issue will be closed once that happens.

Thank you so much!

**Comment by haverchuck** [comment]
@blattr This has been included in version 0.1.6! We will leave this issue open for a while for comment.

**Comment by offlineprogrammer** [comment]
Hi All

I am closing this issue for now due to inactivity 

Regards
Mo
